oauthenticator/oauth2.py

```diff
@@ -1015,6 +1053,19 @@ async def check_allowed(self, username, auth_model):
         if self.allow_all:
             return True
 
+        # If we specific scope grants are required, validate that they have been granted
+        # If not, we explicitly raise an exception here, since they should not be allowed
+        # *regardless* of any other config allowing them access.
+        if self.required_scopes:
+            granted_scopes = auth_model.get('auth_state', {}).get('scope', [])
+            missing_scopes = set(self.required_scopes) - set(granted_scopes)
+            if missing_scopes:
+                message = f"Denying access to user {username}. {self.scope} scopes were requested, {self.required_scopes} are required for authorization, but only {granted_scopes} were granted"
+                self.log.info(message)
+                raise web.HTTPError(403, message)
+            else:
+                return True
```

Comment:
I dom't want required scopes to return true ever, now its line allowing users as long as you have the required scopes.

How then could you use this with any other allow logic? Use of this would then bypass all allow logic - its required, but if you have it, you are authorized and can't distinguish a subset of users meeting this requirement.

Reply:
This is done.